controller/kongplugininstallation/image/image_test.go

```diff
@@ -0,0 +1,120 @@
+package image_test
+
+import (
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"oras.land/oras-go/v2/registry/remote/auth"
+	"oras.land/oras-go/v2/registry/remote/credentials"
+
+	"github.com/kong/gateway-operator/controller/kongplugininstallation/image"
+	"github.com/kong/gateway-operator/test/integration"
+)
+
+func TestCredentialsStoreFromString(t *testing.T) {
+	testCases := []struct {
+		name                string
+		credentials         string
+		expectedErrorMsg    string
+		expectedCredentials func(t *testing.T, cs credentials.Store)
+	}{
+		{
+			name:             "invalid credentials",
+			credentials:      "foo",
+			expectedErrorMsg: "invalid config format:",
+		},
+		{
+			name: "valid credentials",
+			// Field auth is base64 encoded "test:test".
+			credentials: `
+			{
+ 			  "auths": {
+ 			    "ghcr.io": {
+ 			      "auth": "dGVzdDp0ZXN0"
+ 			    }
+ 			  }
+			}`,
+			expectedCredentials: func(t *testing.T, cs credentials.Store) {
+				t.Helper()
+				require.NotNil(t, cs)
+				c, err := cs.Get(context.Background(), "ghcr.io")
+				require.NoError(t, err)
+				require.Equal(t, auth.Credential{Username: "test", Password: "test", RefreshToken: "", AccessToken: ""}, c)
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			credsStore, err := image.CredentialsStoreFromString(tc.credentials)
+			if tc.expectedCredentials != nil {
+				tc.expectedCredentials(t, credsStore)
+			} else {
+				require.ErrorContains(t, err, tc.expectedErrorMsg)
+			}
+		})
+	}
+}
+
+func TestFetchPluginContent(t *testing.T) {
+	t.Run("invalid image URL", func(t *testing.T) {
+		_, err := image.FetchPluginContent(context.Background(), "foo bar", nil)
+		require.ErrorContains(t, err, "unexpected format of image url: could not parse reference: foo bar")
+	})
+
+	const registryUrl = "northamerica-northeast1-docker.pkg.dev/k8s-team-playground/"
```

Comment:
As I understand, this effectively makes this test an integration test (as we're testing against a real registry, with pre-baked images). I think we need to make sure it's properly documented how to access the registry, how to prepare images for testing, etc.

BTW have you considered mocking the registry in tests so that we can inject different images in tests and have the code responsible for creating them close?

Reply:
I can imagine that if we leave this as is and we accidentally remove the remote registry, it will be painful to rebuild it manually without any docs. If there will be docs for that (or a script) for bootstrapping a registry with all the images used in tests, I'm fine with going this way. If not, I think it's better to have some code that will bootstrap it locally (e.g. with [testcontainers](https://github.com/testcontainers/testcontainers-go/blob/main/modules/registry/registry.go) as you mentioned.

Regarding a separate target - I think it's fine to run it as a part of the UT suite as long as we do not need auth credentials to pass.

Reply:
I've created https://github.com/Kong/gateway-operator/issues/465 and added it as a part of the https://github.com/Kong/gateway-operator/issues/371 to do it when the final shape of tests will be reached